syncer/checkpoint_test.go

```diff
@@ -14,29 +14,65 @@
 package syncer
 
 import (
+	"bytes"
 	"fmt"
 	"io/ioutil"
 	"os"
 	"path/filepath"
+	"strings"
 
+	sqlmock "github.com/DATA-DOG/go-sqlmock"
 	. "github.com/pingcap/check"
 	"github.com/siddontang/go-mysql/mysql"
 
 	"github.com/pingcap/dm/dm/config"
+	"github.com/pingcap/dm/pkg/log"
 )
 
-// NOTE: there are binlog events conflict with other test cases
-// and we can not simply disable `sql_log_bin` for this test case
-// because go-sql-driver/mysql using connection pool for *sql.DB
-// and different SQLs in different txn may in different sessions
-// ref: https://github.com/go-sql-driver/mysql/issues/208
-// so disable checkpoint test now, TestCheckPoint => testCheckPoint
-func (s *testSyncerSuite) testCheckPoint(c *C) {
-	id := "test_for_db"
-	cp := NewRemoteCheckPoint(s.cfg, id)
-	defer cp.Close()
-
-	err := cp.Init()
+var (
+	cpid               = "test_for_db"
+	schemaCreateSQL    = ""
+	tableCreateSQL     = ""
+	clearCheckPointSQL = ""
+	loadCheckPointSQL  = ""
+	flushCheckPointSQL = ""
+)
+
+func (s *testSyncerSuite) prepareCheckPointSQL() {
+	schemaCreateSQL = fmt.Sprintf("CREATE SCHEMA IF NOT EXISTS `%s`", s.cfg.MetaSchema)
+	tableCreateSQL = fmt.Sprintf("CREATE TABLE IF NOT EXISTS `%s`.`%s_syncer_checkpoint` .*", s.cfg.MetaSchema, s.cfg.Name)
+	flushCheckPointSQL = fmt.Sprintf("INSERT INTO `%s`.`%s_syncer_checkpoint` .* VALUES.* ON DUPLICATE KEY UPDATE .*", s.cfg.MetaSchema, s.cfg.Name)
+	clearCheckPointSQL = fmt.Sprintf("DELETE FROM `%s`.`%s_syncer_checkpoint` WHERE `id` = '%s'", s.cfg.MetaSchema, s.cfg.Name, cpid)
+	loadCheckPointSQL = fmt.Sprintf("SELECT .* FROM `%s`.`%s_syncer_checkpoint` WHERE `id`='%s'", s.cfg.MetaSchema, s.cfg.Name, cpid)
+}
+
+// this test case uses sqlmock to simulate all SQL operations in tests
+func (s *testSyncerSuite) TestCheckPoint(c *C) {
+	cp := NewRemoteCheckPoint(s.cfg, cpid)
+	defer func() {
+		s.cpMock.ExpectClose()
+		cp.Close()
```

Comment:
can we split this to two `defer`, one for `cp.Close` at here, another for `s.cpMock.ExpectClose()` after L60.
although it may be not a problem for a test.

Reply:
`s.cpMock.ExpectClose()` is safe and does exactly what we want without panic, so I think put them together is ok.